File: collage-gesture-detector/src/main/java/com/cardinalblue/gesture/state/BaseGestureState.java
//  Copyright Oct 2017-present CardinalBlue
//
//  Author: boy@cardinalblue.com
//          jack.huang@cardinalblue.com
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.cardinalblue.gesture.state;

import android.os.Message;
import android.view.MotionEvent;

import com.cardinalblue.gesture.IGestureStateOwner;
import com.cardinalblue.gesture.MyMotionEvent;

public abstract class BaseGestureState {

    protected final IGestureStateOwner mOwner;

    public BaseGestureState(IGestureStateOwner owner) {
        mOwner = owner;
    }

    public abstract void onEnter(MotionEvent event,
                                 Object touchingObject,
                                 Object touchingContext);

    public abstract void onDoing(MotionEvent event,
                                 Object touchingObject,
                                 Object touchingContext);

    public abstract void onExit(MotionEvent event,
                                Object touchingObject,
                                Object touchingContext);

    public abstract boolean onHandleMessage(Message msg);

    ///////////////////////////////////////////////////////////////////////////
    // Protected / Private Methods ////////////////////////////////////////////

    protected MyMotionEvent obtainMyMotionEvent(MotionEvent event) {
        final int action = event.getActionMasked();
        final boolean pointerUp = action == MotionEvent.ACTION_POINTER_UP;
        final int upIndex = pointerUp ? event.getActionIndex() : -1;
        final int downPointerCount = event.getPointerCount() - (pointerUp ? 1 : 0);

        // Prepare the down x and y;
        final float[] downXs = new float[downPointerCount];
        final float[] downYs = new float[downPointerCount];
        for (int i = 0, j = 0; i < event.getPointerCount(); ++i) {
            if (i == upIndex) continue;

            downXs[j] = event.getX(i);
            downYs[j] = event.getY(i);

            ++j;
        }

        if (pointerUp) {
            return new MyMotionEvent(event.getActionMasked(),
                                     downXs, downYs,
                                     true,
                                     event.getX(upIndex),
                                     event.getY(upIndex));

        } else {
            return new MyMotionEvent(event.getActionMasked(),
                                     downXs, downYs,
                                     false, 0, 0);
        }
    }

    protected MyMessagePayload obtainMessagePayload(MotionEvent event,
                                                    Object touchingObject,
                                                    Object touchingContext) {
        MyMotionEvent eventClone = obtainMyMotionEvent(event);

        return new MyMessagePayload(eventClone,
                                    touchingObject,
                                    touchingContext);
    }

    protected Message obtainMessageWithPayload(int what,
                                               MotionEvent event,
                                               Object touchingObject,
                                               Object touchingContext) {
        final Message msg = mOwner.getHandler().obtainMessage(what);
        msg.obj = obtainMessagePayload(event,
                                       touchingObject,
                                       touchingContext);

        return msg;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Clazz //////////////////////////////////////////////////////////////////

    protected static class MyMessagePayload {

        final MyMotionEvent event;
        final Object touchingTarget;
        final Object touchingContext;

        MyMessagePayload(MyMotionEvent event,
                         Object touchingTarget,
                         Object touchingContext) {
            this.event = event;
            this.touchingTarget = touchingTarget;
            this.touchingContext = touchingContext;
        }
    }
}

File: collage-gesture-detector/src/main/java/com/cardinalblue/gesture/state/BaseGestureState.kt
// Copyright Feb 2017-present CardinalBlue
//
// Author: boy@cardinalblue.com
//         jack.huang@cardinalblue.com
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

package com.cardinalblue.gesture.state

import android.os.Message
import android.view.MotionEvent

import com.cardinalblue.gesture.IGestureStateOwner
import com.cardinalblue.gesture.ShadowMotionEvent

abstract class BaseGestureState internal constructor(
    private val mOwner: IGestureStateOwner) {

    protected val owner: IGestureStateOwner get() = mOwner

    abstract fun onEnter(event: MotionEvent,
                         target: Any?,
                         context: Any?)

    abstract fun onDoing(event: MotionEvent,
                         target: Any?,
                         context: Any?)

    abstract fun onExit(event: MotionEvent,
                        target: Any?,
                        context: Any?)

    abstract fun onHandleMessage(msg: Message): Boolean

    ///////////////////////////////////////////////////////////////////////////
    // Protected / Private Methods ////////////////////////////////////////////

    protected fun obtainMyMotionEvent(event: MotionEvent): ShadowMotionEvent {
        val action = event.actionMasked
        val pointerUp = action == MotionEvent.ACTION_POINTER_UP
        val upIndex = if (pointerUp) event.actionIndex else -1
        val downPointerCount = event.pointerCount - if (pointerUp) 1 else 0

        // Prepare the down x and y;
        val downXs = FloatArray(downPointerCount)
        val downYs = FloatArray(downPointerCount)
        var i = 0
        var j = 0
        while (i < event.pointerCount) {
            if (i == upIndex) {
                ++i
                continue
            }

            downXs[j] = event.getX(i)
            downYs[j] = event.getY(i)

            ++j
            ++i
        }

        return if (pointerUp) {
            ShadowMotionEvent(event.actionMasked,
                              downXs, downYs)

        } else {
            ShadowMotionEvent(event.actionMasked,
                              downXs, downYs)
        }
    }

    private fun obtainMessagePayload(event: MotionEvent,
                                     target: Any?,
                                     context: Any?): MyMessagePayload {
        val eventClone = obtainMyMotionEvent(event)

        return MyMessagePayload(eventClone,
                                target,
                                context)
    }

    internal fun obtainMessageWithPayload(what: Int,
                                          event: MotionEvent,
                                          target: Any?,
                                          context: Any?): Message {
        val msg = mOwner.handler.obtainMessage(what)
        msg.obj = obtainMessagePayload(event,
                                       target,
                                       context)

        return msg
    }

    ///////////////////////////////////////////////////////////////////////////
    // Clazz //////////////////////////////////////////////////////////////////

    internal class MyMessagePayload(val event: ShadowMotionEvent,
                                    val target: Any?,
                                    val context: Any?)
}
